### PR TITLE
Add proxy for frontend

### PR DESCRIPTION
**Pull request by kulla**
Fixes https://github.com/serlo/frontend/issues/78

- [x] Fix Search
- [x] cache API responses with KV store
- [x] move constants to environment variables
- [x] API-ENDPOINT als Header weitergeben
- [x] Use Cookie to check route 
- [x] CookieName depends on probability
- [x] Check for entity types to decide which backend should be used
- [x] transfer response and request status code
- [x] setting of cookie (based on probability)
- [x] random decision where to go
- [x] Add route `enable-frontend` and `disable-frontend`
- [x] Prefix `_next` goes to frontend

### CODE REVIEW

**Comment by Entkenntnis** [comment]
Hi Kulla, this implementation looks great!

File: src/frontend-proxy.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+function chooseBackend(
+  req: Request,
+  probability: number
+): { useFrontend: boolean; setCookie: boolean } {
+  const path = getPathname(req.url)
+  const cookies = req.headers.get('Cookie')
+
+  if (path.startsWith('/_next')) return { useFrontend: true, setCookie: false }
```
Comment (Entkenntnis) [inline src/frontend-proxy.ts:39]: https://github.com/serlo/frontend/pull/118 helps to handle all remaining assets, so you can add `if (path.startsWith('/_assets')) ...` here

**Comment by Entkenntnis** [comment]
Fetching the content type fromt the new api:


```tsx
import { request } from 'graphql-request'

const query = alias => `
  {
    uuid (alias:{instance:de, path:"${alias}"}) {
      __typename
    }
  }
`

async function getContentType(alias) {
  const data = await request('https://api.serlo.org/graphql', query(alias))
  return data.uuid.__typename
}
```

File: src/frontend-proxy.ts (hunk 1 above)
Comment (kulla) [reply src/frontend-proxy.ts]: Thanks! I'll implement it...

File: src/frontend-proxy.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
+import { getSubdomain, getPathname } from './url-utils'
+
+export const FRONTEND_DOMAIN = 'frontend-sooty-ten.now.sh'
+const API_ENDPOINT = 'https://api.serlo.org/graphql'
+const FRONTEND_ALLOWED_TYPES = ['Article', 'TaxonomyTerm']
```
Comment (kulla) [inline src/frontend-proxy.ts:5]: @Entkenntnis This variable contains the allowed types for which the frontend can be served. Which types shall I add here?

Comment (Entkenntnis) [reply src/frontend-proxy.ts:5]: Frontend also supports Video, Applet, Course and CoursePage

File: src/frontend-proxy.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  const path = getPathname(url)
+
+  if (getSubdomain(url) !== 'de') return null
+
+  if (path === '/enable-frontend')
+    return createResponse('Enable frontend', probability, true)
+  if (path === '/disable-frontend')
+    return createResponse('Disable frontend', probability, false)
+
+  if (path.startsWith('/_next/') || path.startsWith('/_asset/'))
```
Comment (Entkenntnis) [inline src/frontend-proxy.ts:23]: don't want to nit pick ... but the current prefix is `/_assets/`

**Comment by Entkenntnis** [comment]
for injections and the cookie banner, we would also need to pass through `/api/`

File: src/frontend-proxy.ts (hunk 2 above)
Comment (Entkenntnis) [reply src/frontend-proxy.ts:5]: and landing-page?

**Comment by inyono** [comment]
Search doesn't work

**Comment by inyono** [comment]
https://de.serlo-development.dev/mathe/funktionen/wichtige-funktionstypen-eigenschaften/quadratische-funktionen---parabeln leads to 502

**Comment by inyono** [comment]
To improve performance, we might need to cache the API response at cloudflare worker level (e.g. using Cloudflare Worker KV with a short TTL)

File: __tests__/frontend-proxy.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
+import {
+  handleRequest,
+  FRONTEND_DOMAIN,
+  createApiQuery,
+} from '../src/frontend-proxy'
+import { createJsonResponse } from '../src/utils'
+import { withMockedFetch, hasOkStatus } from './utils'
+
+const TEST_ALLOWED_TYPES = ['User', 'Article']
```
Comment (inyono) [inline __tests__/frontend-proxy.ts:9]: User does conflict with community upgrade, right?

File: __tests__/frontend-proxy.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            const res = await handleUrl(testUrl, probability)
+
+            expect(await res.text()).toBe(responseText)
+            expect(res.headers.get('Set-Cookie')).toBe(
+              `useFrontend${probability * 100}=${useFrontendCookie}; path=/`
+            )
+          }
+        )
+      }
+    )
```
Comment (inyono) [inline __tests__/frontend-proxy.ts:45]: I don't really like this test:
- The `test.each` makes it hard to read. I have to jump between the values, their names and their usage.
- I also have to jump to the top if I want to know what `testUrl` or `TEST_ALLOWED_TYPES` stand for.
- It doesn't feel like checking different types improves the test since the tested assertion here should completely independent of the type, right? If so, I'd argue that you can just choose one type and be done with it (making the test more readable and emphasizing what the actual difference between the test cases is).
- `createApiResponse` & `checkFrontendRequest` are two additional helpers that I need to check to understand this test. `createApiResponse` is probably okay since the naming makes sense, but I don't know what `checkFrontendRequest` is supposed to do, especially as it depends on request and contains logic?
- `withMockedFetch` is another clever helper that I don't really like in the tests. Again, I have to check its implementation (that might be error-prone btw) to understand what it is doing. What does the array represent? Is the order relevant? Why does it accept both responses and functions that create the response? IMHO: inlining the `withMockedFetch` calls does make this way more readable and I only have to understand `jest` to get the gist of what the test wants to achieve.
- Why do we call `handleUrl` when this is another function in the tests that has some backend logic? Why not make the thing that it does explicit and call the production code directly? I shouldn't need to jump to the implementation to see that `handleUrl` apparently also (sometimes?) creates a Cookie before calling the backend logic. That should be part of the test.

What I'd like to strive for in the tests instead is:
- Make it as easy as possible to understand what the tests achieves to check. I want to understand what the test has as assumptions (given), what the tests executes (when) and what it checks (then). In this specific example, given & when are hidden inside `handleUrl` and `withMockedFetch` with some logic behind it (do you assume that the user already has the cookie or not?), assertions are also kinda over the place (in `withMockedFetch` and its callback)
- I'd like to understand what the edge cases are. Did you check `User` and `Article` because they have a different behavior that we need to watch out for? There is no value to checking everything. A test is not a proof but rather a concrete example usage that represents a set of inputs. So I'd rather have one example for the good case + one example each for relevant edge cases. Having more than two tests in the same set of inputs just makes the tests harder to read and increases the time it takes to run the tests.
- Tests should be self-explanatory as possible. If there is too much duplicate code so that we create helpers, helpers should be as "dumb" as possible (no own logic) and restrict themselves to one part of the test (given / when or then). It might even be possible to move common setup into a `beforeEach` and forgo with the helper.

File: __tests__/frontend-proxy.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+  })
+
+  test('creates new response object for calls to frontend', async () => {
+    const frontendResponse = new Response('')
+
+    await withMockedFetch([createApiResponse(), frontendResponse], async () => {
+      const res = await handleUrl(testUrl, 1)
+      expect(res).not.toBe(frontendResponse)
+    })
+  })
```
Comment (inyono) [inline __tests__/frontend-proxy.ts:109]: What's the point of this test?

File: __tests__/frontend-proxy.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+  probability = 0.1,
+  cookie?: string,
+  reqInit?: RequestInit
+) {
+  const req = new Request(url, reqInit)
+
+  if (cookie !== undefined) req.headers.set('Cookie', cookie)
+
+  return (await handleRequest(req, probability, TEST_ALLOWED_TYPES)) as Response
+}
```
Comment (inyono) [inline __tests__/frontend-proxy.ts:195]: Even if you want to use this helper, this should be split into multiple ones. It contains two flag arguments (`cookie` and `reqInit`) that change the assumptions of the test. I'd probably remove the default argument from `probability`, too, since I'd like that to see in the test instead.

File: __tests__/frontend-proxy.ts [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+  })
+}
+
+function checkFrontendRequest(req: Request) {
+  const res =
+    new URL(req.url).hostname === FRONTEND_DOMAIN
+      ? new Response('frontend')
+      : new Response('no-frontend')
+
+  return new Promise<Response>((resolve) => resolve(res))
```
Comment (inyono) [inline __tests__/frontend-proxy.ts:214]:
```suggestion
  return Promise.resolve(res)
```

File: __tests__/frontend-proxy.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    errors: [{ message: 'error' }],
+    data: { uuid: null },
+  })
+}
+
+function checkFrontendRequest(req: Request) {
+  const res =
+    new URL(req.url).hostname === FRONTEND_DOMAIN
+      ? new Response('frontend')
+      : new Response('no-frontend')
```
Comment (inyono) [inline __tests__/frontend-proxy.ts:212]: Not really sure what this is supposed to do. Is this a stub for the actual response that might either from serlo.org or frontend?

File: __tests__/maintenance.ts [hunk 10]
```diff
@@ -21,27 +21,18 @@
...
   test('Disabled (no maintenance planned)', async () => {
     mockMaintenanceKV({})
-    await handleRequest('https://de.serlo.org')
-    expectFetchToHaveBeenCalledWithRequest({
-      url: 'https://de.serlo.org',
-    } as Request)
+
+    await withMockedFetch([createApiResponse(), echoUrl], async () => {
+      const res = await handleUrl('https://de.serlo.org')
+      expect(await res.text()).toBe('https://de.serlo.org/')
```
Comment (inyono) [inline __tests__/maintenance.ts:34]: Why does the maintenance need to know anything about the API? Either Maintenance is in front of the proxy logic (i.e. when maintenance mode is on, we won't call the proxy) or behind (i.e. there is only a maintenance mode for non-frontend).

Comment (inyono) [reply __tests__/maintenance.ts:34]: And I don't like the refactor with mockedFetch. Previously, the code showed exactly what it did. Now we have some weird assertion on `res.text()` which is a fake response created via `echoUrl`...

File: src/frontend-proxy.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
+import { getSubdomain, getPathname } from './url-utils'
+
+export const FRONTEND_DOMAIN = 'frontend-sooty-ten.now.sh'
+const API_ENDPOINT = 'https://api.serlo.org/graphql'
```
Comment (inyono) [inline src/frontend-proxy.ts:4]: Should be dependent on the environment so we can test out a new API version. You might even think about sending that down to the frontend as a header so that the frontend uses a different API for different environments.

File: src/frontend-proxy.ts (hunk 2 above)
Comment (inyono) [reply src/frontend-proxy.ts:5]: IMHO: we should white-list the types one-by-one instead of everything at once (and integrate feedback on the way)

File: src/frontend-proxy.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+const FRONTEND_ALLOWED_TYPES = ['Article', 'TaxonomyTerm']
+const FRONTEND_PROBABILITY = 0.1
+
+export async function handleRequest(
+  request: Request,
+  probability = FRONTEND_PROBABILITY,
+  allowdTypes = FRONTEND_ALLOWED_TYPES
```
Comment (inyono) [inline src/frontend-proxy.ts:11]: typo

File: src/frontend-proxy.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  if (path.startsWith('/_next/') || path.startsWith('/_asset/'))
+    return await fetchBackend(true, false)
+
+  const typename = await queryTypename(path)
+  if (typename === null || !allowdTypes.includes(typename)) return null
+
+  const cookies = request.headers.get('Cookie')
+
+  if (cookies?.includes(formatCookie(true, probability)))
+    return fetchBackend(true, false)
```
Comment (inyono) [inline src/frontend-proxy.ts:32]: probably more readable if you combine that into an object `{useFrontend: true, setCookie: false}`

File: src/frontend-proxy.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+export function createApiQuery(path: string): string {
+  const pathWithoutSlash = path.startsWith('/') ? path.slice(1) : path
+  const query = /^\/?\d+$/.test(pathWithoutSlash)
+    ? `id: ${pathWithoutSlash}`
+    : `alias: { instance: de, path: "/${pathWithoutSlash}" }`
+
+  return `{ uuid(${query}) { __typename } }`
+}
+
+async function createResponse(
```
Comment (inyono) [inline src/frontend-proxy.ts:75]: Moe that into `handleRequest` so that it's near the location it is called

File: src/frontend-proxy.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  resInfo: string | Response,
+  probability: number,
+  futureFrontendUse?: boolean
+) {
+  const body = typeof resInfo === 'string' ? resInfo : await resInfo.text()
+  const init = typeof resInfo === 'string' ? undefined : resInfo
+  const response = new Response(body, init)
```
Comment (inyono) [inline src/frontend-proxy.ts:82]: should be two different functions, `createResponseFromBody` and `createResponseFromRequest` or something like that (flag argument)

File: src/frontend-proxy.ts [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+  if (futureFrontendUse !== undefined) {
+    const cookie = `${formatCookie(futureFrontendUse, probability)}; path=/`
+
+    response.headers.set('Set-Cookie', cookie)
```
Comment (inyono) [inline src/frontend-proxy.ts:87]: Could be its own function, doesn't need to be called by `createResponse`

File: src/frontend-proxy.ts [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+  if (path === '/disable-frontend')
+    return createResponse('Disable frontend', probability, false)
+
+  if (path.startsWith('/_next/') || path.startsWith('/_assets/'))
+    return await fetchBackend(true, false)
+
+  if (path !== '/') {
+    const typename = await queryTypename(path)
+    if (typename === null || !allowdTypes.includes(typename)) return null
+  }
```
Comment (kulla) [inline src/frontend-proxy.ts:37]: @Entkenntnis With `landing-page` you meant that a request to `/` (main page) should also be able to be forwarded to frontend or did you meant other landing pages as well?

Comment (Entkenntnis) [reply src/frontend-proxy.ts:37]: Jup, I meant /

File: __tests__/frontend-proxy.ts [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+import { createJsonResponse } from '../src/utils'
+import { withMockedFetch, hasOkStatus } from './utils'
+
+const TEST_ALLOWED_TYPES = ['User', 'Article']
+const TEST_URLS = [
+  'https://de.serlo.org/math',
+  'https://de.serlo.org/10',
+  'https://de.serlo.org/_nextexample-url',
+]
+
```
Comment (kulla) [inline __tests__/frontend-proxy.ts:15]: camelCase for constants

**Comment by kulla** [comment]
> To improve performance, we might need to cache the API response at cloudflare worker level (e.g. using Cloudflare Worker KV with a short TTL)

- [ ] catch typename with KV-store only if typename exists (1h)

File: __tests__/frontend-proxy.ts (hunk 5 above)
Comment (inyono) [reply __tests__/frontend-proxy.ts]:
```
const fetch = jest.fn().mockReturnValueOnce(value)
global.fetch = fetch
fetch.mockImplementationOnce(createApiResponse(typename))

const res = await handleUrl(testUrl, probability)
expectFetchToHaveBeenCalledWithRequest({
    url: 'frontend.serlo.org', // Or whatever,
    headers: {
        'Set-Cookie': `useFrontend${probability * 100}=${useFrontendCookie}; path=/`
    }
})
```

Comment (kulla) [reply __tests__/frontend-proxy.ts]: - [ ] more 2 argumente -> object argument
- [ ] delete TEST_URLS and TEST_ALLOWED_TYPES

File: src/frontend-proxy.ts (hunk 11 above)
Comment (kulla) [reply src/frontend-proxy.ts:4]: - [ ] API-ENDPOINT als Header weitergeben

File: src/frontend-proxy.ts [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+  return `{ uuid(${query}) { __typename } }`
+}
+
+async function createResponse(
+  resInfo: string | Response,
+  probability: number,
+  futureFrontendUse?: boolean
+) {
+  const body = typeof resInfo === 'string' ? resInfo : await resInfo.text()
```
Comment (Entkenntnis) [inline src/frontend-proxy.ts:90]: maybe need to handle resInfo.blob() for images and fonts

Comment (inyono) [reply src/frontend-proxy.ts:90]: Mh... that's definitely weird since it works correctly for assets.serlo.org. Maybe check the proxy there again and see if there are any differences to the frontend proxy

**Comment by kulla** [comment]
> Search doesn't work

Seems to be fixed now by another fix of mine. Can you double check?

**Comment by kulla** [comment]
@inyono I also changed the code so that the calculated random number is saved in the cookie. Thus someone who already get the new design continues to get it when we increase the probability for the frontend backend.

File: src/frontend-proxy.ts [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+  if (path === '/disable-frontend') {
+    const response = new Response('Disable frontend')
+
+    setCookieUseFrontend(response, 1)
+
+    return response
+  }
+
+  if (path.startsWith('/_next/') || path.startsWith('/_assets/'))
```
Comment (Entkenntnis) [inline src/frontend-proxy.ts:30]: Just one small addition: `/api/` is necessary for injections and other client-side request to work, could you add it please? Thanks!

**Comment by kulla** [comment]
@inyono This PR is ready to be merged.

File: src/vars.d.ts [hunk 21]
```diff
@@ -19,5 +19,24 @@
...
  */
-declare let DOMAIN: string
-declare let ENABLE_BASIC_AUTH: 'true' | 'false'
+declare namespace NodeJS {
+  interface Global {
+    API_ENDPOINT: string
+    DOMAIN: string
+    ENABLE_BASIC_AUTH: 'true' | 'false'
+    FRONTEND_ALLOWED_TYPES: string
+    FRONTEND_CACHE_TYPES: KV<string>
```
Comment (inyono) [inline src/vars.d.ts:28]: For KVs, there is already `kv.d.ts`

File: src/vars.d.ts [hunk 22]
```diff
@@ -19,5 +19,24 @@
...
+    FRONTEND_DOMAIN: string
+    FRONTEND_PROBABILITY: string
+  }
+}
+
+declare
+interface KV<ValueType> {
```
Comment (inyono) [inline src/vars.d.ts:35]: ValueType is always string

File: wrangler.toml [hunk 23]
```diff
@@ -19,8 +27,16 @@ route = "*serlo-staging.dev/*"
 kv-namespaces = [
     { binding = "MAINTENANCE_KV", id = "3c3c25aba0b94b7e99f855f013f5b09c" },
     { binding = "PACKAGES_KV", id = "19f90dc8e6ff49cd8bc42f51346409be" },
+    { binding = "FRONTEND_CACHE_TYPES", id="a323aa5ec11a4940abe7fb5ab5d95a39" },
 ]
-vars = { DOMAIN = "serlo-staging.dev", ENABLE_BASIC_AUTH = "true" }
+
+[env.staging.vars]
+API_ENDPOINT = "https://api.serlo.org/graphql"
```
Comment (inyono) [inline wrangler.toml:34]:
```suggestion
API_ENDPOINT = "https://api.serlo-staging.dev/graphql"
```

File: wrangler.toml [hunk 24]
```diff
@@ -9,8 +9,16 @@ route = "*serlo-development.dev/*"
 kv-namespaces = [
     { binding = "MAINTENANCE_KV", id = "d5f281b897f643b8824881cfe2a24dd2" },
     { binding = "PACKAGES_KV", id = "19f90dc8e6ff49cd8bc42f51346409be" },
+    { binding = "FRONTEND_CACHE_TYPES", id="802462c937f24f268bb523fa855ee4a1" },
 ]
-vars = { DOMAIN = "serlo-development.dev", ENABLE_BASIC_AUTH = "true" }
+
+[env.dev.vars]
+API_ENDPOINT = "https://api.serlo.org/graphql"
```
Comment (inyono) [inline wrangler.toml:16]:
```suggestion
API_ENDPOINT = "https://api.serlo-development.dev/graphql"
```

**Comment by inyono** [comment]
Suggested a couple of things (still not too happy about the tests), only thing that needs to be fixed before releasing this is https://github.com/serlo/serlo.org-cloudflare-worker/pull/11#discussion_r421207611

**Comment by inyono** [comment]
Another blocking issue: if the user is logged in, they should always get the legacy backend because the frontend pretty much misses every feature for authorized users

**Comment by kulla** [comment]
@inyono Blocking issues for merge are implemented. See https://github.com/serlo/serlo.org-cloudflare-worker/pull/11/commits/158db2b103bb6616e72f7b82d0f9025d35859ca7 and https://github.com/serlo/serlo.org-cloudflare-worker/pull/11/commits/76aeb8c005c33232c2b4a69e88b554f4a360d35b for the changes. #13 for the follow up issues...

File: src/frontend-proxy.ts [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+  }: {
+    useFrontend: number
+    setCookie: boolean
+  }) {
+    const backendUrl =
+      useFrontend <= probability
+        ? `https://${global.FRONTEND_DOMAIN}${getPathname(request.url)}`
+        : request.url
+    const backendRequest = new Request(backendUrl, request)
+    backendRequest.headers.set('X-SERLO-API', global.API_ENDPOINT)
```
[the remark this answers is not in the document]
Comment (kulla) [reply src/frontend-proxy.ts:63]: See https://github.com/serlo/frontend/issues/177

**Comment by kulla** [comment]
@inyono Today I also implemented the remaining feedback :-)
